### PR TITLE
Add offloading tests and fix obscure edge case

### PR DESCRIPTION
**Pull request by janeyx99**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [x] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Adds basic unit test for offloading for correctness
* Fixes a really obscure bug that could happen (it has never been reported yet) and adds a test for that

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/training/_activation_offloading.py [hunk 1]
```diff
@@ -281,8 +285,21 @@ def wait_and_del_remaining_references() -> None:
...
-                        event = self.s0.record_event()
-                        self.bwd_ev_stash[unpack_tensor_id] = event
+                        # if any of the outputs is a view of the tensor, meaning the tensor might be used later,
+                        # we cannot presume to delete it after only the current node is done! So we use our frenemy,
+                        # record_stream, to ensure the Tensor stays unmessed with until it's done getting used
+                        # in the compute stream (s0 here). Note that the con here is we introduce non-deterministic
+                        # memory usage, but this case should not happen often.
+                        unpacked_tensor = self.bwd_tensor_stash[unpack_tensor_id]
+                        if any(
+                            o.untyped_storage() is unpacked_tensor.untyped_storage()
```
Comment (ringohoffman) [inline torchtune/training/_activation_offloading.py:295]: When I run with this, I get:

```
    o.untyped_storage() is unpacked_tensor.untyped_storage()
AttributeError: 'NoneType' object has no attribute 'untyped_storage'
```

I think `o` _can_ be None.

Comment (janeyx99) [reply torchtune/training/_activation_offloading.py:295]: Good catch thanks! Should be updated in latest commit

**Review by albanD — COMMENTED** [review]
SGTM!

**Review by joecummings — APPROVED** [review]
applause
